src/hotspot/share/gc/shared/oopStorage.cpp

```diff
@@ -999,8 +984,8 @@ bool OopStorage::delete_empty_blocks() {
     }
   }
   // Exceeded work limit or can't delete last block.  This will
-  // cause the service thread to loop, giving other subtasks an
-  // opportunity to run too.  There's no need for a notification,
+  // cause the ServiceThread to loop, giving other subtasks an
+  // opportunity to run too. There's no need for a notification,
```

Comment:
With the changes to `has_cleanup_work_and_reset` this no longer causes the ServiceThread to loop.
Instead it requests cleanup at the next scheduled time for the ServiceThread to do so.  And there's no
longer ever any notification, so the final sentence needs some adjustment.

Reply:
Hm, with the change to `has_cleanup_work_and_reset` this will result in the service thread deleting
no more than "work limit" blocks per "defer period".  Maybe this should reset the "permit time" too,
so that it _does_ cause the ServiceThread to loop.

Reply:
The ServiceThread loops because it has a fixed wait timeout now.  It won't process these every time it's notified which can be more frequent.  This comment seems to say that there's more work, so do it next time.  I guess resetting the permit time to zero would prevent these cleanups from getting behind.  Are there tests for this condition?

Reply:
Never mind. I misremembered how the work limiting operated. It's not a fixed
limit on how much work to do. Rather, it's (roughly) process at most the
number of blocks as there were in the list on entry. The point is that if
other threads are allocating and then emptying blocks while we're working,
that can't cause us to keep working for some potentially arbitrary amount of
time.

Also, the result indicating more work to do is unused by the ServiceThread.
It is used by the gtest for delete_empty_blocks, but I think will never be
true as used there.

We could change ServiceThread to pay attention to the result, but that would
make that code a bit more complicated, and because of how the "work limit"
works I don't think there's much benefit to that.  There would be if the "work
limit" were some arbitrary fixed count, like 10 blocks or something, but since
it's not...

So I think just further updating the comment is sufficient.  I think just keep
the first sentence ("Exceeded ... last block.") and delete the rest, about
making the ServiceThread loop.

I might file a new RFE to do something useful with that bool result or
eliminate it.

Reply:
But we still want to have record_needs_cleanup() right?  So that the ServiceThread will find work to do on the next iteration, but it's true that it wont *cause* the service thread to loop.

Reply:
Yes, or at least not immediately loop, because of the deferral period.